### ISSUE TITLE
Git Clone on Non-Standard SSH Port Hangs

### ISSUE THREAD

**Issue by ratmav** [ansible/ansible-modules-core#1886]
##### Issue Type:

Bug Report
##### Ansible Version:

ansible 2.0.0 (devel e16b24c01e) last updated 2015/08/06 13:20:23 (GMT -400)
  lib/ansible/modules/core: (detached HEAD 4c622a0338) last updated 2015/08/06 13:20:32 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD 210ee3febc) last updated 2015/08/06 13:20:32 (GMT -400)
  v1/ansible/modules/core: (detached HEAD f8d8af17cd) last updated 2015/08/04 17:12:56 (GMT -400)
  v1/ansible/modules/extras: (detached HEAD 495ad450e5) last updated 2015/08/04 17:13:01 (GMT -400)
  config file = /home/ratmav/ansible.cfg
  configured module search path = None
##### Ansible Configuration:

Added the following the the default config, pulled from [here](https://raw.githubusercontent.com/ansible/ansible/devel/examples/ansible.cfg). Disabled cows.

``` bash
# Forward agent for Git clones, etc.
ssh_args = -o ForwardAgent=yes
```
##### Environment:
- Local workstation is Ubuntu 14.04 LTS
- Remote host is also Ubuntu 14.04 LTS
##### Summary:

Git module hangs on clone when using a non-standard SSH port. I'd like to be able to clone successfully from something other than 22 on my remote origin.
##### Steps To Reproduce:

_Assumes remote origin with SSH access on a port other than 22, local Ansible user's SSH key is authorized for the Git user on the remote origin._

``` bash
- name: clone project
  git:
    repo: 'ssh://git@remote.origin.ip.address:PORT/home/git/repositories/project.git'
    dest: ~/project
    version: "{{ git_branch }}"
    accept_hostkey: true
```
##### Expected Results:

Clone completes successfully; source files are copied to remote host.
##### Actual Results:

Ansible hangs indefintely, no error. Running `-vvvv`, this is as far as I get (scrubbed):

``` bash
<remote.host.ip.address> ESTABLISH SSH CONNECTION FOR USER: deployment-user

<remote.host.ip.address> EXEC ssh -C -tt -vvv -o ForwardAgent=yes -o StrictHostKeyChecking=no -o Port=NON-STANDARD SSH PORT -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=deployment-user -o ConnectTimeout=10 remote.host.ip.address mkdir -p "$HOME/.ansible/tmp/ansible-tmp-1438882538.44-99189208925421" && chmod a+rx "$HOME/.ansible/tmp/ansible-tmp-1438882538.44-99189208925421" && echo "$HOME/.ansible/tmp/ansible-tmp-1438882538.44-99189208925421"

<remote.host.ip.address> ESTABLISH SSH CONNECTION FOR USER: deployment-user

<remote.host.ip.address> PUT /tmp/tmpGF10B8 TO /home/deployment-user/.ansible/tmp/ansible-tmp-1438882538.44-99189208925421/git

<remote.host.ip.address> ESTABLISH SSH CONNECTION FOR USER: deployment-user

<remote.host.ip.address> EXEC ssh -C -tt -vvv -o ForwardAgent=yes -o StrictHostKeyChecking=no -o Port=NON-STANDARD SSH PORT -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=deployment-user -o ConnectTimeout=10 remote.host.ip.address LANG=C LC_MESSAGES=C LC_CTYPE=C /usr/bin/python /home/deployment-user/.ansible/tmp/ansible-tmp-1438882538.44-99189208925421/git; rm -rf "/home/deployment-user/.ansible/tmp/ansible-tmp-1438882538.44-99189208925421/" >/dev/null 2>&1
```

The above output sits indefinitely, so I `<CTRL>+c` out to get the typical wrap up message. I never see an error:

``` bash
PLAY RECAP **********************************************************************
remote-host           : ok=2    changed=0    unreachable=0    failed=0
```
## Details:

The clone runs fine after setting SSHD on the remote origin to use port 22 and removing the port from the task:

``` bash
- name: clone project
  git:
    repo: 'ssh://git@remote.origin.ip.address/home/git/repositories/project.git'
    dest: ~/project
    version: "{{ git_branch }}"
    accept_hostkey: true
```

I've also tried:

``` bash
- name: clone project
  git: repo=ssh://git@remote.origin.ip.address:PORT/home/git/repositories/project.git dest=~/project...etc.
```

I feel like this may be related to https://github.com/ansible/ansible/issues/7941, but the fix for that was merged into the main repo a year ago.

**Comment by bcoca** [comment]
Can You Help Us Out?
====================

Thanks for filing a ticket! I am the friendly GitHub Ansibot.

It looks like you might not have filled out the issue description based on our standard issue template.  You might not have known about that, and that's ok too, we'll tell you how to do it.

We have a standard template because Ansible is a really busy project and it helps to have some standard information in each ticket, and GitHub doesn't yet provide a standard facility to do this like some other bug trackers.  We hope you understand as this is really valuable to us!.

Solving this is simple: please copy the contents of this [template](https://raw.githubusercontent.com/ansible/ansible/devel/ISSUE_TEMPLATE.md) and **paste it into the description** of your ticket.  That's it!

If You Had A Question To Ask Instead
====================================

If you happened to have a "how do I do this in Ansible" type of question, that's probably more of a user-list question than a bug report, and you should probably ask this question on the project mailing list instead.

However, if you think you have a bug, the report is the way to go!  We definitely want all the bugs filed :)  Just trying to help!

About Priority Tags
===================

Since you're here, we'll also share some useful information at this time.

In general tickets will be assigned a priority between P1 (highest) and P5, and then worked in priority order. We may also have some follow up questions along the way, so keeping up with follow up comments via GitHub notifications is a good idea.

Due to large interest in Ansible, humans may not comment on your ticket immediately.

Mailing Lists
=============

If you have concerns or questions, you're welcome to stop by the ansible-project or ansible-development mailing lists, as appropriate.  Here are the links:

  * https://groups.google.com/forum/#!forum/ansible-project - for discussion of bugs and how-to type questions
  * https://groups.google.com/forum/#!forum/ansible-devel - for discussion on how to implement a code change, or feature brainstorming among developers

Thanks again for the interest in Ansible!

**Comment by ratmav** [comment]
Ok, I did a recent pull, etc., and now Ansible is throwing the following:

``` bash
fatal: [remote_host]: FAILED! => {"changed": false, "cmd": "/usr/bin/git ls-remote origin -h refs/heads/master", "failed": true, "msg": "ssh: connect to host remote.host.ip.address 
port 22: Connection refused\r\nfatal: Could not read from remote repository.\n\nPlease make sure you have the correct access rights\nand the repository exists.", "rc": 128, "std
err": "ssh: connect to host remote.host.ip.address port 22: Connection refused\r\nfatal: Could not read from remote repository.\n\nPlease make sure you have the correct access rights\nan
d the repository exists.\n", "stdout": "", "stdout_lines": []}
```

The task again, for reference:

``` bash
- name: clone project
  git:
    repo: ssh://git@remote.host.ip.address:PORT/home/git/repositories/project.git
    dest: ~/project
    version: "{{ git_branch }}"
    accept_hostkey: true
```

Changing the remote origin's SSH port to 22 and removing the port from the Git SSH connection string in the task clones successfully. It looks like Ansible is just dropping the set port and using 22 as a rule.

**Comment by sivel** [comment]
I cannot confirm this behavior and have tested by running the following:

```
git:
    repo: ssh://git@localhost:1234/sivel/repo.git
    dest: /tmp/repo
    accept_hostkey: yes
    version: master
```

The command executed by ansible is:

```
/usr/local/bin/git ls-remote ssh://git@localhost:1234/sivel/repo.git -h refs/heads/master
```

And the module completes successfully with:

```
{"changed": true, "after": "8322ede13aef8090dde913d1d33464df1eb6660a", "before": null}
```

**Comment by ratmav** [comment]
OK, I'll try running that command locally in a terminal and see what spills out. Hopefully it's just something broken on my end rather than some edge case.

**Comment by ratmav** [comment]
I tracked this down earlier today: It's my fault [according to the git module docs](http://docs.ansible.com/ansible/git_module.html), and it exposes some interesting (or new to me) behavior:

I ran the above command via the `shell` module, and the clone ran fine. Good news. So it has to be my something subtle in my task. Here's the original task:

``` yaml
- name: clone project
  git:
    repo: ssh://git@remote.host.ip.addy:PORT/home/git/repositories/project.git
    dest: ~/project
    version: "{{ git_branch }}"
    accept_hostkey: true
```

Here's the working task:

``` yaml
- name: clone project
  git:
    repo: ssh://git@remote.host.ip.addy:PORT/home/git/repositories/project.git
    dest: ~/project
    version: "{{ git_branch }}"
    accept_hostkey: yes # HERE
```

So, I transposed `yes` and `true` in my task for the `accept_hostkey` parameter, which would explain the hang (waiting on key acceptance). The interesting part is that the task doesn't fail loudly, it hangs, and everything appears to work fine when SSH is on port 22. It might be worth either having stricter enforcement on the `accept_hostkey` value so it will fail loudly or perhaps just accepting `true` as `yes`.

I'm not sure if ya'll want to close the issue at this point as user error, or perhaps look at this as an edge case issue with user input. Either/or works for me; I appreciate all the help.

**Comment by ratmav** [comment]
Ok, so something is going awry on my end with `become`/`become_user` and `accept_hostkey` as well:

Works:

``` yaml
- name: clone project
  git:
    repo: ssh://git@remote.host.ip.addy:PORT/home/git/repositories/project.git
    dest: ~/project
    version: "{{ git_branch }}"
    accept_hostkey: yes # HERE
```

Hangs:

``` yaml
- name: clone project
  git:
    repo: ssh://git@remote.host.ip.addy:PORT/home/git/repositories/project.git
    dest: ~/project
    version: "{{ git_branch }}"
    accept_hostkey: yes # HERE
  become: yes
  become_user: deployment_user
```

Can you reproduce this?

**Comment by ratmav** [comment]
#1343 looks pretty similar to this.

**Comment by ratmav** [comment]
Did a pull today, clone looks like it's working. Not sure whether the breakage was on my side or Ansible's, but I'm going to assume it's my problem.

**Comment by jp1987** [comment]
Exact same situation doesn't work for me either, on Ansible 2.2.0. It works without become/become_user, but hangs when I add that.

**Comment by iceteabottle** [comment]
I've found a simple solution:
Just add these lines to git config file on the target machine `~/.ssh/config` 
```
Host=git.domain.com
Hostname=git.domain.com
Port=20022
```
Then the git clone works without the explicit port.

Btw. it's a normal git thing. This doesn't work either `git clone git@git.domain.com:20022/some-repo`. With the config above it will work without the port `git clone git@git.domain.com:some-repo`
